Determine a flex-items used flex-basis

dlrobertson opened the pull request:
Determine a given flex-items flex-basis, use the given flex-basis as the hypothetical main size, and size the item into the available space with this size.

I've been working on this for quite some time, but this is still very much a work in progress. This PR dramiatically changes the way we size flex items. We currently ignore the size and stretch items to fit [evenly in the available space](https://github.com/servo/servo/blob/master/components/layout/flex.rs#L248) and ignore their size. As a result, I suspect a lot of tests that have been passing will fail with this PR.

Comments and critiques would be very welcome!

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/10945)

<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/values.rs

danlrobertson commented:
Review status: 0 of 11 files reviewed at latest revision, 7 unresolved discussions.

---

*[components/layout/flex.rs, line 185 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10945#-KGcYBzXNLRttL-TQtXs:-KGcYBzXNLRttL-TQtXt:1815073541) ([raw file](https://github.com/servo/servo/blob/68456ccc419e8914e73b3bfea114795756b356f3/components/layout/flex.rs#L185)):*
I'm sure there is a better way to do this. Comments would be appreciated.

---

*[components/layout/flex.rs, line 301 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10945#-KGcYLHuW7VuU8c-a1_A:-KGcYLHuW7VuU8c-a1_B:-1550403847) ([raw file](https://github.com/servo/servo/blob/68456ccc419e8914e73b3bfea114795756b356f3/components/layout/flex.rs#L301)):*
This doesn't "override" the items width. I couldn't figure out how to do this. As a result the following tests will fail.
```
tests/wpt/css-tests/css-flexbox-1_dev/html/flex-basis-005.htm
tests/wpt/css-tests/css-flexbox-1_dev/html/flex-basis-006.htm
```

---

*[components/layout/flex.rs, line 445 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10945#-KGcYte1-qib_xaRsvkW:-KGcYte2zTaNo0_yAZJr:-911826052) ([raw file](https://github.com/servo/servo/blob/68456ccc419e8914e73b3bfea114795756b356f3/components/layout/flex.rs#L445)):*
I realized that in my prior PR this was placed in the wrong location

---

*[components/layout/flex.rs, line 449 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10945#-KGcZ-Z-UYFkQTIHHzFg:-KGcZ-Z0lTq0SiXStdwk:1854746597) ([raw file](https://github.com/servo/servo/blob/68456ccc419e8914e73b3bfea114795756b356f3/components/layout/flex.rs#L449)):*
Same here

---

*[components/style/values.rs, line 1081 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10945#-KGcZ2oloinaeviIodAF:-KGcZ2omy_NYdtjrqpSt:-864164235) ([raw file](https://github.com/servo/servo/blob/68456ccc419e8914e73b3bfea114795756b356f3/components/style/values.rs#L1081)):*
This was added to get `tests/wpt/css-tests/css-flexbox-1_dev/html/flexbox_computedstyle_flex-basis-0percent.htm` to pass. I couldn't find the portion of the spec for this. Please double check me on this.

---

*[tests/wpt/metadata-css/css-flexbox-1_dev/html/flex-basis-005.htm.ini, line 3 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10945#-KGcZUotTveR1vJ-yetI:-KGcZUouAVr7aXXPydwT:-1709293226) ([raw file](https://github.com/servo/servo/blob/26be403e3c8351af638cce3a2a1ed4d3be8022df/tests/wpt/metadata-css/css-flexbox-1_dev/html/flex-basis-005.htm.ini#L3)):*
Currently fails. See the comment above.

---

*[tests/wpt/metadata-css/css-flexbox-1_dev/html/flex-basis-006.htm.ini, line 3 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10945#-KGcZYaUnIHxmuRqP0IH:-KGcZYaUnIHxmuRqP0II:-1709293226) ([raw file](https://github.com/servo/servo/blob/26be403e3c8351af638cce3a2a1ed4d3be8022df/tests/wpt/metadata-css/css-flexbox-1_dev/html/flex-basis-006.htm.ini#L3)):*
Currently fails. See the comment above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10945)*
<!-- Sent from Reviewable.io -->

danlrobertson commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 68456cc with merge c9ade33...

bors-servo commented:
:broken_heart: Test failed - [windows](http://build.servo.org/builders/windows/builds/360)

danlrobertson commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit b959416 with merge d59b161...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/678)

highfive commented:
▶ FAIL [expected PASS] /css-flexbox-1_dev/html/flex-001.htm
      └   → /css-flexbox-1_dev/html/flex-001.htm 7c51a382f0afafc8cd4b2473bb60f333a0807cae
    /css-flexbox-1_dev/html/reference/justify-content-001-ref.htm b9206fc45cf7be2854714e1d32797e49d61013aa
    Testing 7c51a382f0afafc8cd4b2473bb60f333a0807cae == b9206fc45cf7be2854714e1d32797e49d61013aa
    
      ▶ FAIL [expected PASS] /css-flexbox-1_dev/html/flex-basis-005.htm
      └   → /css-flexbox-1_dev/html/flex-basis-005.htm 7d53c98773aae0945c22d7e5bbe05999f7ccd0c7
    /css-flexbox-1_dev/html/reference/ref-filled-green-100px-square.htm 82d94d2027e216f7a2d31fcc53afb5e048516cef
    Testing 7d53c98773aae0945c22d7e5bbe05999f7ccd0c7 == 82d94d2027e216f7a2d31fcc53afb5e048516cef
    
      ▶ FAIL [expected PASS] /css-flexbox-1_dev/html/flex-basis-006.htm
      └   → /css-flexbox-1_dev/html/flex-basis-006.htm 7d53c98773aae0945c22d7e5bbe05999f7ccd0c7
    /css-flexbox-1_dev/html/reference/ref-filled-green-100px-square.htm 82d94d2027e216f7a2d31fcc53afb5e048516cef
    Testing 7d53c98773aae0945c22d7e5bbe05999f7ccd0c7 == 82d94d2027e216f7a2d31fcc53afb5e048516cef
    
      ▶ FAIL [expected PASS] /css-flexbox-1_dev/html/flex-flexitem-percentage-prescation.htm
      └   → /css-flexbox-1_dev/html/flex-flexitem-percentage-prescation.htm 3f84cc7ddb35700085b908cc4d149b53ba46f08e
    /css-flexbox-1_dev/html/reference/flex-flexitem-percentage-prescation-ref.htm 03fce235f75accc82c7f5487464a47a0294caa9d
    Testing 3f84cc7ddb35700085b908cc4d149b53ba46f08e == 03fce235f75accc82c7f5487464a47a0294caa9d
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flex-grow-003.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flex-grow-</span><span class="stdout">005.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flex-shrink-004.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flex-shrink-005.htm
    
      ▶ FAIL [expected PASS] /css-flexbox-1_dev/html/flexbox-align-self-stretch-vert-001.htm
      └   → /css-flexbox-1_dev/html/flexbox-align-self-stretch-vert-001.htm 44fb1820e47487538c00d39c63fb1d897bbdd40b
    /css-flexbox-1_dev/html/reference/flexbox-align-self-stretch-vert-001-ref.htm 79533a8e47a74d7ac166d7e4a21f8195f7bc3926
    Testing 44fb1820e47487538c00d39c63fb1d897bbdd40b == 79533a8e47a74d7ac166d7e4a21f8195f7bc3926
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flexbox-anonymous-items-001.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flexbox-flex-wrap-default.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flexbox-flex-wrap-nowrap.htm
    
      ▶ FAIL [expected PASS] /css-flexbox-1_dev/html/flexbox_align-items-stretch-writing-modes.htm
      └   → /css-flexbox-1_dev/html/flexbox_align-items-stretch-writing-modes.htm 5b9ff21634cf640aa9b8394b40ccf7d59bf5feab
    /css-flexbox-1_dev/html/reference/flexbox_align-items-stretch-writing-modes-ref.htm 6245bcdba8c46ab7d46bef2c52881f2c432ebad1
    Testing 5b9ff21634cf640aa9b8394b40ccf7d59bf5feab == 6245bcdba8c46ab7d46bef2c52881f2c432ebad1
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flexbox_stf-abspos.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flexbox_stf-float.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flexbox_stf-inline-block.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flexbox_stf-table-cell.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flexbox_stf-table-row-group.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flexbox_stf-table-row.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flexbox_stf-table.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flexbox_table-fixed-layout.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/flexible-box-float.htm
    
      ▶ PASS [expected FAIL] /css-flexbox-1_dev/html/justify-content-002.htm

danlrobertson commented:
Review status: 0 of 11 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[components/layout/flex.rs, line 301 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10945#-KGcYLHuW7VuU8c-a1_A:-KGsX8ADl7uKfKmo0fUk:-1024967333) ([raw file](https://github.com/servo/servo/blob/b9594168e663ff1fb2fcd0ce967159443ec15c6a/components/layout/flex.rs#L301)):*
I have since realized that this comment is wrong. This isn't a problem of overriding the items width, but rather the [distribution of free space](https://drafts.csswg.org/css-flexbox/#algo-main-align)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10945)*
<!-- Sent from Reviewable.io -->

dlrobertson committed:
Determine a flex-items flex-basis

Determine a given flex-items flex-basis and use the given flex-basis as
the hypothetical main size.

highfive commented:
New code was committed to pull request.

stshine commented:
I think the terminology 'flex basis' specifies the initial main size of a flex item. Maybe another name for your usage?

danlrobertson commented:
@stshine: Thanks for the comment! What part in particular are you referring to? This particular PR is attempting to start work on the [following portion of the spec](https://drafts.csswg.org/css-flexbox/#algo-main-item).

stshine commented:
@danlrobertson I mean that  the flex-basis is uniquely determined by the style, so call the function flex_base_size() maybe is better.

stshine commented:
Review status: 0 of 11 files reviewed at latest revision, 8 unresolved discussions.

---

*[components/layout/flex.rs, line 173 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10945#-KGyVcFGQhrMI6Qenwpd:-KGyVcFGQhrMI6Qenwpe:397794922) ([raw file](https://github.com/servo/servo/blob/b9594168e663ff1fb2fcd0ce967159443ec15c6a/components/layout/flex.rs#L173)):*
I believe the minimum size of a flex item is not related to flex-basis, but to the 'min-width' property.

---

*[components/layout/flex.rs, line 185 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10945#-KGcYBzXNLRttL-TQtXs:-KGyT0HbqYMfJp-rIM3U:-1407693562) ([raw file](https://github.com/servo/servo/blob/b9594168e663ff1fb2fcd0ce967159443ec15c6a/components/layout/flex.rs#L185)):*
What interesting here is that both webkit and gecko do not respect flex-basis of flex items when computing the intrinsic main size of a flex container. However, to implement this properly, the algorithm in https://drafts.csswg.org/css-flexbox-1/#intrinsic-sizes seems to be more complicated. For now, I suggest use computation.union_inline() if the container can be wrapped. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10945)*
<!-- Sent from Reviewable.io -->

danlrobertson commented:
Gotcha. Yeah, that makes sense

---

Review status: 0 of 11 files reviewed at latest revision, 8 unresolved discussions.

---

*[components/layout/flex.rs, line 173 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10945#-KGyVcFGQhrMI6Qenwpd:-KGyZQKTZTLk9VVSOqec:-971904725) ([raw file](https://github.com/servo/servo/blob/b9594168e663ff1fb2fcd0ce967159443ec15c6a/components/layout/flex.rs#L173)):*
Ah, didn't know that. Thanks for the tip! So this should actually implement https://drafts.csswg.org/css-flexbox-1/#intrinsic-sizes vs https://drafts.csswg.org/css-flexbox/#algo-main-item here?

---

*[components/layout/flex.rs, line 185 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10945#-KGcYBzXNLRttL-TQtXs:-KGyj8IZbUe9Z0o-5pj7:1864283417) ([raw file](https://github.com/servo/servo/blob/b9594168e663ff1fb2fcd0ce967159443ec15c6a/components/layout/flex.rs#L185)):*
Hmmm... We currently don't support wrapped lines. That is next on my TODO list. Any thoughts on what we should do when the container can't be wrapped?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10945#-:-KGyZMRKn9OBhmM-2a7k:1968991560)*
<!-- Sent from Reviewable.io -->

stshine commented:
Review status: 0 of 11 files reviewed at latest revision, 8 unresolved discussions.

---

*[components/layout/flex.rs, line 173 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10945#-KGyVcFGQhrMI6Qenwpd:-KGypIebcmsjkpjvPl3w:2009318355) ([raw file](https://github.com/servo/servo/blob/b9594168e663ff1fb2fcd0ce967159443ec15c6a/components/layout/flex.rs#L173)):*
Yes, this function bubbles the intrinsic inline size of the flex flow to its parent.

---

*[components/layout/flex.rs, line 185 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10945#-KGcYBzXNLRttL-TQtXs:-KGypVN8P88KzraXMScS:2076778712) ([raw file](https://github.com/servo/servo/blob/b9594168e663ff1fb2fcd0ce967159443ec15c6a/components/layout/flex.rs#L185)):*
There is a lot to do,and the most significant one is that when these is only one line and there cross size of a flex container is definite and has property align-items:stretch, all flex item should be considered as having definite cross size. But we should implement all the flex properties first.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10945)*
<!-- Sent from Reviewable.io -->

stshine commented:
May I ask how is the progressing on the implemention of flexbox? I would like to make some contribution on it.

danlrobertson commented:
Definitely! That would be very welcome. I'm sure you already know this, but the main file is `components/layout/flex.rs`. Feel free to contact me on `#servo` at `dlr`. I'll try to dig up some issues soon.

---

Review status: 0 of 11 files reviewed at latest revision, 8 unresolved discussions.

---

*[components/layout/flex.rs, line 173 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10945#-KGyVcFGQhrMI6Qenwpd:-KH5f6GLbWVmJf0zIb7v:1162598759) ([raw file](https://github.com/servo/servo/blob/b9594168e663ff1fb2fcd0ce967159443ec15c6a/components/layout/flex.rs#L173)):*
Awesome. I'll be uploading a large update to this PR this weekend. The one question this leaves me with is that this is the first pass of reflow (bottom up), so we don't have the containing block size. So if the current item is sized with a percentage is `wrap` wouldn't we need to know the size of the container? Or since the size is definite would we just skip this part since `fixed_width == true`?

---

*[components/layout/flex.rs, line 185 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10945#-KGcYBzXNLRttL-TQtXs:-KH5fkP1GQUzp1X8p-07:-2100479889) ([raw file](https://github.com/servo/servo/blob/b9594168e663ff1fb2fcd0ce967159443ec15c6a/components/layout/flex.rs#L185)):*
Ah true. I noted that [here](https://drafts.csswg.org/css-flexbox/#definite-sizes). There is definitely a lot left to be implemented

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10945#-:-KH5eRtEYRDNqqfrSYiT:799738690)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #10987) made this pull request unmergeable. Please resolve the merge conflicts.

stshine commented:
Hi, Sorry for the late. I have been investigating into this topic for the past week,
1. the container size is not known during the bubble_inline_sizes() pass, so this function simply can't be used at that phase.
2. What makes the situation more complex, is that if the container is vertical and the flex-basis is content based, then the item has to be reflowed first under 'fit-content' width to get the flex base size; what's more, if if has min-height:auto property,  it may needs to be reflowed again under 'max-content' width to get min-height size. I decided to work on horizontal flex container first and functions to get flex base size, outer hypothetical main size, minimal main size and max main size.

<details><summary><i>Previously, bors-servo wrote…</i></summary>
> :umbrella: The latest upstream changes (presumably #10987) made this pull request unmergeable. Please resolve the merge conflicts.</details>


---

Review status: 0 of 11 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[components/layout/flex.rs, line 173 \[r1\]](https://reviewable.io:443/reviews/servo/servo/10945#-KGyVcFGQhrMI6Qenwpd:-KHr_XwxMbhUw7JVwgxi:1092691028) ([raw file](https://github.com/servo/servo/blob/b9594168e663ff1fb2fcd0ce967159443ec15c6a/components/layout/flex.rs#L173)):*
<details><summary><i>Previously, danlrobertson (Dan Robertson) wrote…</i></summary>
> Awesome. I'll be uploading a large update to this PR this weekend. The one question this leaves me with is that this is the first pass of reflow (bottom up), so we don't have the containing block size. So if the current item is sized with a percentage is `wrap` wouldn't we need to know the size of the container? Or since the size is definite would we just skip this part since `fixed_width == true`?</details>
It is skiped in the current implemention.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10945#-:-KHlrh-bc8pTFIqfEFGC:-1192192448)*
<!-- Sent from Reviewable.io -->

danlrobertson commented:
> 1. the container size is not known during the bubble_inline_sizes() pass, so this function simply can't be used at that phase.

Yeah, I realized that after some `gdb`. Must be why I put it there in the first place :smile:.

> flex-basis is content based, then the item has to be reflowed first under 'fit-content' width to get the flex base size; 

Yes, the spec mentions.

> If a cross size is needed to determine the main size (e.g. when the flex item’s main size is in its block axis) and the flex item’s cross size is auto and not definite

Awesome! Sorry for my late reply as well. I'm in the middle of starting a new job and a "hack week". But life should be getting back to normal here, so I should be able to focus more on this now.

danlrobertson commented:
As stated on slack, I think the vast majority of this can be replaced by the newer work by @stshine

stshine commented:
Currently In my opinion the `AxisSize` struct still can find its usage, just that it should be used by flex items rather than flex container. For example, in a vertical flex container the main size in the vertical axis have a distinction of be indefinite or definite, even through they all have a definite value. While for now I am busy with figuring out the correct way of setting inline sizes of a flex item, which are determined by the flex container, not by itself like the how it is done in `block.rs`, I will reply when I find the correct usage of this PR.

stshine commented:
Hi @danlrobertson, I decided to stop using `AxisSize` and use the `MaybeAuto` type in model.rs instead, since 1. this type comes with some handy helper methods, which is common used by other modules in layout to represent size of a flow, and using it would help  the interoperability between flex flow and other flows, 2. The `MaybeAuto`, `Option<Au>`, and `AxisSize` without `MinMaxConstraint` is very similar to each other, and remove one can help decrease redundancy, 3. For now I have not been able to  found a tight relationship between `MinMaxConstraint` and the size.

On an other hand, I have improved the `determine_flex_basis()` function to make it accept the `flex_basis` and return a value of  `MaybeAuto` type to represent the used flex base size. And the `MinMaxConstraint` may find its usage in the future since in a lot of place servo do not clamp size as the spec specified.

Please consider close this pull request. If you hope this PR to be merged, please rewrite the `determine_flex_basis()` function to make it accept the flex basis and return the `MaybeAuto` type.

I plan to make some PRs recently, sorry for the late (God the float related logic is nightmare).
